numpy/core/src/npysort/quicksort.c.src

```diff
@@ -83,11 +85,21 @@
  *         npy_uint, npy_long, npy_ulong, npy_longlong, npy_ulonglong,
  *         npy_ushort, npy_float, npy_double, npy_longdouble, npy_cfloat,
  *         npy_cdouble, npy_clongdouble, npy_datetime, npy_timedelta#
+ * #AVX512 = 0*5, 1, 1, 0*5, 1, 0*7#
  */
 
 NPY_NO_EXPORT int
 quicksort_@suff@(void *start, npy_intp num, void *NOT_USED)
 {
+#if @AVX512@
+#if defined HAVE_ATTRIBUTE_TARGET_AVX512_SKX_WITH_INTRINSICS
+    if(NPY_CPU_HAVE(AVX512_SKX)) {
+        avx512_qsort_@suff@(start, num);
+        return 0;
+    }
+#endif
+#endif
+
```

Comment:
@seiko2plus you mean this should be removed and the code should use the new dispatcher semantics?

Reply:
Yes but first we need a dispatch-able source for the avx512_skx kernels or x86 features in general. in other words, `qsort-32bit-avx512.h.src` should be renamed to `x86-qsort.dispatch.c.src`. also, we need another header or any other existence header to hold the forward declarations of the exported functions. for example:

> 1. `x86-qsort.h` (forward declarations)

```C
#include "npy_cpu_dispatch.h"

#ifndef NPY_DISABLE_OPTIMIZATION
    #include "x86-qsort.dispatch.h"
#endif
NPY_CPU_DISPATCH_DECLARE(NPY_NO_EXPORT void x86_quicksort_int,
     (void *start, npy_intp num))

NPY_CPU_DISPATCH_DECLARE(NPY_NO_EXPORT void x86_quicksort_uint,
     (void *start, npy_intp num))
```

> 2.`x86-qsort.dispatch.c.src` (kernels)

```C
/*@targets
 * $maxopt $keep_baseline avx512_skx
 */
// policy $keep_baseline is used to avoid skip building avx512_skx
// when its part of baseline features (--cpu-baseline), since
// 'baseline' option isn't specified within targets.

#include "x86-qsort.h"

#ifdef NPY_HAVE_AVX512_SKX
NPY_NO_EXPORT void NPY_CPU_DISPATCH_CURFX(x86_quicksort_int)
(void *start, npy_intp num)
{
    // avx512 simd code for unsinged int
}

NPY_NO_EXPORT void NPY_CPU_DISPATCH_CURFX(x86_quicksort_uint)
(void *start, npy_intp num)
{
    // avx512 simd code for singed int
}
#endif
```
> 3. runtime dispatch

```diff
+#include "x86-qsort.h"
-#if defined -HAVE_ATTRIBUTE_TARGET_AVX512_SKX_WITH_INTRINSICS
-    if(NPY_CPU_HAVE(AVX512_SKX)) {
-        avx512_qsort_@suff@(start, num);
-        return 0;
-    }
-#endif
+#ifndef NPY_DISABLE_OPTIMIZATION
+    #include "x86-qsort.dispatch.h"
+#endif
+void (*dispfunc)(void*, npy_intp) = NULL;
+NPY_CPU_DISPATCH_CALL_XB(dispfunc = &x86_quicksort_@suff@);
+if (dispfunc) {
+   (*dispfunc)(start, num);
+   return 0;
+}
```
NOTE: `NPY_CPU_DISPATCH_CALL_XB` is no longer accept goto or return since #17201, while makes it easier for readability but as u can realize I had to add extra unnecessary lines to get the 'return' back (compiler going to optimize it out anyway).
before #17201, the runtime-dispatch code can be simplified as follows:
```C
#ifndef NPY_DISABLE_OPTIMIZATION
    #include "x86-qsort.dispatch.h"
#endif
NPY_CPU_DISPATCH_CALL_XB(x86_quicksort_@suff@, (start, num), return 0);
```

Reply:
Thanks for spelling it out :) Would have found it hard to figure out on my own.